jvm/core/src/main/scala/org/scalatest/tools/SocketReporter.scala

```diff
@@ -15,28 +15,50 @@
  */
 package org.scalatest.tools
 
-import org.scalatest.events.Event
+import org.scalatest.events._
 import org.scalatest.ResourcefulReporter
 import java.net.Socket
 import java.io.ObjectOutputStream
 import java.io.BufferedOutputStream
+import java.util.concurrent.atomic.AtomicReference
 
 private[scalatest] class SocketReporter(host: String, port: Int) extends ResourcefulReporter {
 
-  private val socket = new Socket(host, port)
-  private val out = new ObjectOutputStream(socket.getOutputStream)
+  private val socket = new AtomicReference(new Socket(host, port))
+  private val out = new AtomicReference(new ObjectOutputStream(socket.get.getOutputStream))
+
+  def refresh(): Unit = {
+    try {
+      out.get.close()
+      socket.get.close()
+    }
+    catch {
+      case _: Throwable =>
+    }
+    socket.set(new Socket(host, port))
+    out.set(new ObjectOutputStream(socket.get.getOutputStream))
+  }
   
   def apply(event: Event): Unit = {
     synchronized {
-      out.writeObject(event)
-      out.flush()
+      try {
+        out.get.writeObject(event)
+        out.get.flush()
+      }
+      catch {
+        case e: java.io.NotSerializableException =>
+          refresh()
+          out.get.writeObject(event.ensureSerializable())
+
+        case e: Throwable => refresh()
+      }
```

Comment:
@cheeseng Here's another catch of Throwable. If you want to handle most exceptions uniformly, then I'd suggest Exception on Scala's NonFatal extractor.